File: lua/stack/02_evaluate_reverse_polish_notation.lua
OPERATORS = {}

function OPERATORS:new()
    local tbl = {}
    setmetatable(tbl, self)
    self.__index = self
    return tbl
end

function OPERATORS:add(left, right)
    return left + right
end

function OPERATORS:substract(left, right)
    return left - right
end

function OPERATORS:multiply(left, right)
    return left * right
end

function OPERATORS:divide(left, right)
    return left / right
end

function OPERATORS:exponentiation(left, right)
    return left ^ right
end

function OPERATORS:modulo(left, right)
    return left % right
end

function OPERATORS:is_operator(token)
    local valid_tokens = { '+', '-', '*', '/', '^', '%' }
    for index = 1, #valid_tokens do
        if valid_tokens[index] == token then
            return true
        end
    end
    return false
end

OPERATION = OPERATORS:new()

function OPERATION:new(args)
    local tbl = {}
    setmetatable(tbl, self)
    self.__index = self
    tbl.tokens = args
    return tbl
end

function OPERATION:calculate()
    local stack = {}
    for index = 1, #self.tokens do
        local token = self.tokens[index]
        if tonumber(token) then
            table.insert(stack, tonumber(token))
        elseif self:is_operator(token) then
            local right = table.remove(stack)
            local left = table.remove(stack)
            local result = nil
            if token == '+' then
                result = self:add(left, right)
            elseif token == '-' then
                result = self:substract(left, right)
            elseif token == '*' then
                result = self:multiply(left, right)
            elseif token == '/' then
                result = self:divide(left, right)
            elseif token == '^' then
                result = self:exponentiation(left, right)
            elseif token == '%' then
                result = self:modulo(left, right)
            end
            table.insert(stack, result);
        end
    end
    return stack[1]
end

-- the garbage collector is supposed to act by itself if a value become unreachable.
local operation = OPERATION:new({ "3", "4", "+", "2", "*" })
print(operation:calculate())
operation = OPERATION:new({"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"})
print(operation:calculate())
operation = OPERATION:new({ "3", "2", "+", "4", "*" })
print(operation:calculate())
operation = OPERATION:new({ "5", "4", "7", "+", "*" })
print(operation:calculate())

-- Setting a value for nil can possibly help the garbage collection operations. 
-- seems like good practice anyway...
function OPERATION:cleanup()
    self.tokens = nil
end

operation:cleanup()

RPN = {}

function RPN:call(tokens)
    local rpn = OPERATION:new(tokens)
    local result = rpn:calculate()
    rpn:cleanup()
    return result
end

setmetatable(RPN, {
    __call = function(self, tokens)
        return self:call(tokens)
    end
})

print("as a function call")

print(RPN({ "5", "4", "7", "+", "*" }))
print(RPN({ "3", "2", "+", "4", "*" }))
print(RPN({ "10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+" }))

File: rust/stack/02_evaluate_reverse_polish_notation.rs

trait Arithmetics {
    fn add(left:i32, right:i32) -> i32;
    fn sub(left:i32, right:i32) -> i32;
    fn multiply(left:i32, right:i32) -> i32;
    fn divide(left:i32, right:i32) -> i32;
    fn modulo(left:i32, right:i32) -> i32;
    fn exponent(left:i32, right:i32) -> i32;
    fn is_operator(token:String) -> bool;
    fn is_number(s: String) -> bool;
}

struct Operators;

impl Arithmetics for Operators {

    fn add(left:i32, right:i32) -> i32 {
        return left + right;
    }
    fn sub(left:i32, right:i32) -> i32 {
        return left - right;
    }
    fn multiply(left:i32, right:i32) -> i32 {
        return left * right;
    }
    fn divide(left:i32, right:i32) -> i32 {
        return left / right;
    }
    fn modulo(left:i32, right:i32) -> i32 {
        return left % right;
    }
    fn exponent(left:i32, right:i32) -> i32 {
        return left ^ right;
    }
    fn is_operator(token:String) -> bool {
        let valid:[&str; 6] = ["+", "-", "*", "/", "^", "%"];
        for index in 0..6 {
            if valid[index] == token {
                return true;
            }
        }
    return false;
    }
    fn is_number(s: String) -> bool {
        s.parse::<i32>().is_ok()
    }
}

struct Operation {
    tokens:Vec<String>,
}

impl Operation {
    pub fn new(input:Vec<&str>) -> Self {
        Self {
            tokens: input
                .iter()
                .map(|s| s.to_string())
                .collect(),
        }
    }
    fn calculate(&self) -> i32 {
        let mut stack:Vec<i32> = Vec::new();
        for index in 0..self.tokens.len() {
            if Operators::is_number(self.tokens[index].clone()) {
                if let Ok(num) = self.tokens[index].parse::<i32>() {
                    stack.push(num);
                }
            } else if Operators::is_operator(self.tokens[index].clone()) {
                let right = stack.pop();
                let left = stack.pop();
                if let Some((left, right)) = left.zip(right) {
                    let result = match self.tokens[index].as_str() {
                        "+" => Operators::add(left, right),
                        "-" => Operators::sub(left, right),
                        "*" => Operators::multiply(left, right),
                        "/" => Operators::divide(left, right),
                        "%" => Operators::modulo(left, right),
                        "^" => Operators::exponent(left, right),
                        _ => 0,
                    };
                    stack.push(result);
                }
            }
        }
        stack[0]
    }
}

fn main() {
    {
        let input:Vec<&str> = vec!["1", "2", "+", "3", "*", "4"];       
    let operation = Operation::new(input);
    println!("{}",operation.calculate());
    }
    {
        let input:Vec<&str> = vec!["3", "4", "+", "2", "*"];          
    let operation = Operation::new(input);
    println!("{}",operation.calculate());
    }
    {
        let input:Vec<&str> = vec!["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"];
    let operation = Operation::new(input);
    println!("{}",operation.calculate());
    }
    {
        let input:Vec<&str> = vec![ "3", "2", "+", "4", "*" ];          
    let operation = Operation::new(input);
    println!("{}",operation.calculate());
    }
}